cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -162,13 +164,43 @@
     return metaTable.executeUnchecked(new TransactionExecutor.Function<DatasetContext<Table>, List<ArtifactDetail>>() {
       @Override
       public List<ArtifactDetail> apply(DatasetContext<Table> context) throws Exception {
-        List<ArtifactDetail> archives = Lists.newArrayList();
+        List<ArtifactDetail> artifacts = Lists.newArrayList();
         Scanner scanner = context.get().scan(scanArtifacts(namespace));
         Row row;
         while ((row = scanner.next()) != null) {
-          addArchivesToList(archives, row);
+          addArchivesToList(artifacts, row);
         }
-        return Collections.unmodifiableList(archives);
+        return Collections.unmodifiableList(artifacts);
+      }
+    });
+  }
+
+  /**
+   * Get all artifacts that match artifacts in the given ranges.
+   *
+   * @param ranges the ranges to match artifacts in
+   * @return an unmodifiable list of all artifacts that match the given ranges. If none exist, an empty list
+   *         is returned
+   */
+  public List<ArtifactDetail> getArtifacts(final Set<ArtifactRange> ranges) {
+    return metaTable.executeUnchecked(new TransactionExecutor.Function<DatasetContext<Table>, List<ArtifactDetail>>() {
+      @Override
+      public List<ArtifactDetail> apply(DatasetContext<Table> context) throws Exception {
+        List<ArtifactDetail> artifacts = Lists.newArrayList();
+        Table table = context.get();
+        for (ArtifactRange range : ranges) {
+          ArtifactKey artifactKey = new ArtifactKey(range.getNamespace(), range.getName());
```

Comment:
good catch, forgot to add the version check.

this is also kind of a weird api, going to change it to take a single range instead of a set.